File: frontend/src/lib/tasks/cloud-tasks.ts
/**
 * Google Cloud Tasks client wrapper
 * Handles enqueueing async jobs to the AI service
 */

let CloudTasksClient: any
let client: any = null

try {
  const tasks = require('@google-cloud/tasks')
  CloudTasksClient = tasks.CloudTasksClient
  const cloudTasksEnabled = process.env.ENABLE_CLOUD_TASKS === 'true'
  client = cloudTasksEnabled ? new CloudTasksClient() : null
  if (!cloudTasksEnabled) {
    console.log('Cloud Tasks disabled, using direct HTTP calls')
  }
} catch (error) {
  console.warn('Cloud Tasks not available, using direct HTTP calls')
}

const PROJECT_ID = process.env.GCS_PROJECT_ID || 'studybuddy-dev'
const LOCATION = process.env.CLOUD_TASKS_LOCATION || 'us-central1'
const QUEUE = process.env.CLOUD_TASKS_QUEUE || 'studybuddy-jobs'
const AI_SERVICE_URL = process.env.AI_SERVICE_URL || 'http://localhost:8000'
const parsePositiveInt = (value: string | undefined, fallback: number): number => {
  const parsed = Number.parseInt(String(value || ''), 10)
  if (Number.isFinite(parsed) && parsed > 0) return parsed
  return fallback
}
const DIRECT_TASK_TIMEOUT_MS = parsePositiveInt(process.env.DIRECT_TASK_TIMEOUT_MS, 10 * 60 * 1000)
const DIRECT_TASK_MAX_RETRIES = parsePositiveInt(process.env.DIRECT_TASK_MAX_RETRIES, 3)
const NON_RETRYABLE_STATUSES = new Set([400, 401, 403, 404, 422])
const INLINE_DIRECT_JOB_TYPES = new Set(
  String(process.env.DIRECT_INLINE_JOB_TYPES || '')
    .split(',')
    .map((entry) => entry.trim())
    .filter(Boolean)
)

export interface TaskPayload {
  jobId: string
  jobType: string
  data: Record<string, any>
}

export class TaskEnqueueError extends Error {
  retryable: boolean
  status?: number

  constructor(message: string, retryable = true, status?: number) {
    super(message)
    this.name = 'TaskEnqueueError'
    this.retryable = retryable
    this.status = status
  }
}

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms))

function isPermanentTaskFailure(status: number, message: string): boolean {
  if (NON_RETRYABLE_STATUSES.has(status)) {
    return true
  }
  const normalized = message.toLowerCase()
  return (
    normalized.includes('openai_api_key is not configured') ||
    normalized.includes('missing bearer') ||
    normalized.includes('authentication') ||
    normalized.includes('invalid api key')
  )
}

async function callAiServiceWithRetry(endpoint: string, payload: TaskPayload): Promise<void> {
  let attempt = 0
  let lastError: TaskEnqueueError | null = null

  while (attempt < DIRECT_TASK_MAX_RETRIES) {
    attempt += 1
    try {
      const response = await fetch(`${AI_SERVICE_URL}${endpoint}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(payload),
        signal: AbortSignal.timeout(DIRECT_TASK_TIMEOUT_MS),
      })

      if (!response.ok) {
        const message = await response.text()
        const retryable = !isPermanentTaskFailure(response.status, message)
        throw new TaskEnqueueError(
          `[DEV] Task execution failed (${response.status}): ${message}`,
          retryable,
          response.status
        )
      }
      return
    } catch (error) {
      const normalized = error instanceof TaskEnqueueError
        ? error
        : new TaskEnqueueError(String(error), true)
      lastError = normalized

      if (!normalized.retryable || attempt >= DIRECT_TASK_MAX_RETRIES) {
        break
      }

      const backoff = 250 * (2 ** (attempt - 1)) + Math.floor(Math.random() * 120)
      await sleep(backoff)
    }
  }

  throw lastError || new TaskEnqueueError('Unknown task enqueue failure')
}

async function markJobFailed(
  payload: TaskPayload,
  error: TaskEnqueueError
): Promise<void> {
  try {
    const { prisma } = await import('@/lib/db/prisma')
    await prisma.processingJob.updateMany({
      where: {
        id: payload.jobId,
        status: { in: ['pending', 'processing'] },
      },
      data: {
        status: 'failed',
        errorCode: error.retryable ? 'ENQUEUE_FAILED' : 'ENQUEUE_PERMANENT',
        errorMessage: error.message.slice(0, 1000),
        retryable: error.retryable,
        completedAt: new Date(),
      },
    })
  } catch (dbError) {
    console.error('[DEV] Failed to persist enqueue failure:', dbError)
  }
}

async function dispatchDirectTask(endpoint: string, payload: TaskPayload): Promise<void> {
  try {
    await callAiServiceWithRetry(endpoint, payload)
  } catch (error) {
    const normalized = error instanceof TaskEnqueueError
      ? error
      : new TaskEnqueueError(String(error), true)
    console.error(`[DEV] Background task dispatch failed for ${payload.jobId}:`, normalized.message)
    await markJobFailed(payload, normalized)
  }
}

async function runInlineDirectTask(endpoint: string, payload: TaskPayload): Promise<void> {
  try {
    await callAiServiceWithRetry(endpoint, payload)
  } catch (error) {
    const normalized = error instanceof TaskEnqueueError
      ? error
      : new TaskEnqueueError(String(error), true)
    await markJobFailed(payload, normalized)
    throw normalized
  }
}

if (!client && INLINE_DIRECT_JOB_TYPES.size > 0) {
  console.warn(
    `[DEV] Inline direct tasks enabled for: ${Array.from(INLINE_DIRECT_JOB_TYPES).join(', ')}`
  )
}

/**
 * Enqueue a task to Cloud Tasks
 * @param endpoint - AI service endpoint (e.g., /jobs/validate-material)
 * @param payload - Task payload
 * @returns Task name
 */
export async function enqueueTask(
  endpoint: string,
  payload: TaskPayload
): Promise<string> {
  if (!client) {
    if (INLINE_DIRECT_JOB_TYPES.has(payload.jobType)) {
      // For user-driven short tasks, run inline in direct mode to reduce queue lag.
      console.log(`[DEV] Running inline task ${payload.jobType} (${payload.jobId})`)
      await runInlineDirectTask(endpoint, payload)
    } else {
      // Heavy/background tasks remain non-blocking.
      console.log(`[DEV] Enqueueing background task ${payload.jobType} (${payload.jobId})`)
      queueMicrotask(() => {
        void dispatchDirectTask(endpoint, payload)
      })
    }

    return `dev-task-${payload.jobId}`
  }

  // Production mode - use Cloud Tasks
  const parent = client.queuePath(PROJECT_ID, LOCATION, QUEUE)

  const task = {
    httpRequest: {
      httpMethod: 'POST' as const,
      url: `${AI_SERVICE_URL}${endpoint}`,
      headers: {
        'Content-Type': 'application/json',
      },
      body: Buffer.from(JSON.stringify(payload)).toString('base64'),
    },
  }

  const [response] = await client.createTask({ parent, task })
  return response.name || `task-${payload.jobId}`
}

/**
 * Enqueue a material validation job
 * @param jobId - Processing job ID
 * @param materialId - Material ID to validate
 * @returns Task name
 */
export async function enqueueValidationJob(
  jobId: string,
  materialId: string
): Promise<string> {
  return enqueueTask('/jobs/validate-material', {
    jobId,
    jobType: 'validate_material',
    data: { materialId },
  })
}

/**
 * Enqueue a topic extraction job
 * @param jobId - Processing job ID
 * @param projectId - Project ID
 * @returns Task name
 */
export async function enqueueTopicExtractionJob(
  jobId: string,
  projectId: string
): Promise<string> {
  return enqueueTask('/jobs/extract-topics', {
    jobId,
    jobType: 'extract_topics',
    data: { projectId },
  })
}

/**
 * Enqueue a content generation job
 * @param jobId - Processing job ID
 * @param topicId - Topic ID
 * @param contentType - Type of content to generate
 * @param preferences - Generation preferences
 * @returns Task name
 */
export async function enqueueContentGenerationJob(
  jobId: string,
  topicId: string,
  contentType: string,
  preferences: Record<string, any>
): Promise<string> {
  return enqueueTask('/jobs/generate-content', {
    jobId,
    jobType: 'generate_content',
    data: { topicId, contentType, preferences },
  })
}

/**
 * Enqueue a chunking job
 * @param jobId - Processing job ID
 * @param materialId - Material ID to chunk
 * @returns Task name
 */
export async function enqueueChunkingJob(
  jobId: string,
  materialId: string
): Promise<string> {
  return enqueueTask('/jobs/chunk-material', {
    jobId,
    jobType: 'chunk_material',
    data: { materialId },
  })
}

/**
 * Enqueue an exam grading job
 * @param jobId - Processing job ID
 * @param submissionId - Exam submission ID
 * @param examId - Exam ID
 * @param questions - Exam questions
 * @param answers - Student answers
 * @returns Task name
 */
export async function enqueueExamGradingJob(
  jobId: string,
  submissionId: string,
  examId: string,
  questions: any[],
  answers: Record<string, any>
): Promise<string> {
  return enqueueTask('/jobs/grade-exam', {
    jobId,
    jobType: 'grade_exam',
    data: { submissionId, examId, questions, answers },
  })
}

File: ai-service/app/services/cloud_tasks.py
"""Google Cloud Tasks client wrapper for AI service.

Handles enqueueing async jobs.
"""
import logging
import json
import base64
from typing import Dict, Any, Optional
import httpx
from app.config import settings

logger = logging.getLogger(__name__)

# Initialize Cloud Tasks client for production
_client: Optional[Any] = None

if settings.is_production:
    try:
        from google.cloud import tasks_v2
        _client = tasks_v2.CloudTasksClient()
    except Exception as e:
        logger.warning(f"Failed to initialize Cloud Tasks client: {e}")
        _client = None


async def enqueue_task(
    endpoint: str,
    payload: Dict[str, Any]
) -> str:
    """
    Enqueue a task to Cloud Tasks or make direct HTTP call in development.

    Args:
        endpoint: AI service endpoint (e.g., /jobs/chunk-material)
        payload: Task payload with jobId, jobType, and data

    Returns:
        Task name or task ID
    """
    if settings.is_development:
        # Development mode - make direct HTTP call
        logger.info(f"[DEV] Triggering job via direct HTTP call to {endpoint}")

        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    f"{settings.AI_SERVICE_URL}{endpoint}",
                    json=payload,
                    timeout=30.0
                )

                if response.status_code != 200:
                    logger.error(f"[DEV] Task execution failed: {response.text}")
                else:
                    logger.info(f"[DEV] Task triggered successfully")

        except Exception as e:
            logger.error(f"[DEV] Failed to call AI service: {e}")

        return f"dev-task-{payload['jobId']}"

    # Production mode - use Cloud Tasks
    if not _client:
        logger.error("Cloud Tasks client not available in production")
        return f"error-task-{payload['jobId']}"

    try:
        # Build the task
        project_id = settings.GCS_PROJECT_ID
        location = "us-central1"  # Same as deploy.yml
        queue = "studybuddy-jobs"  # Same as deploy.yml

        parent = _client.queue_path(project_id, location, queue)

        task = {
            "http_request": {
                "http_method": tasks_v2.HttpMethod.POST,
                "url": f"{settings.AI_SERVICE_URL}{endpoint}",
                "headers": {
                    "Content-Type": "application/json",
                },
                "body": base64.b64encode(json.dumps(payload).encode()).decode(),
            }
        }

        response = _client.create_task(request={"parent": parent, "task": task})
        logger.info(f"Created Cloud Task: {response.name}")
        return response.name

    except Exception as e:
        logger.error(f"Failed to create Cloud Task: {e}")
        return f"error-task-{payload['jobId']}"


async def enqueue_chunking_job(job_id: str, material_id: str) -> str:
    """
    Enqueue a material chunking job.

    Args:
        job_id: Processing job ID
        material_id: Material ID to chunk

    Returns:
        Task name
    """
    return await enqueue_task(
        "/jobs/chunk-material",
        {
            "jobId": job_id,
            "jobType": "chunk_material",
            "data": {"materialId": material_id}
        }
    )


async def enqueue_topic_extraction_job(job_id: str, project_id: str) -> str:
    """
    Enqueue a topic extraction job.

    Args:
        job_id: Processing job ID
        project_id: Project ID

    Returns:
        Task name
    """
    return await enqueue_task(
        "/jobs/extract-topics",
        {
            "jobId": job_id,
            "jobType": "extract_topics",
            "data": {"projectId": project_id}
        }
    )
